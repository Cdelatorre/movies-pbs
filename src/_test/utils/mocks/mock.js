export const moviesData = {
  viewedMovies: [
    {
      id: "1",
      title: "Movie 1",
      genres: ["Horror", "Comedy"],
      viewed: true,
      img: "https://picsum.photos/id/118/300/400",
      createdAt: "2022-10-16T20:25:58.420Z",
    },
  ],
  unViewedMovies: [
    {
      id: "2",
      title: "Movie 2",
      genres: ["Horror", "Comedy"],
      viewed: false,
      img: "https://picsum.photos/id/159/300/400",
      createdAt: "2022-10-16T20:51:29.081Z",
    },
    {
      id: "3",
      title: "Movie 3",
      genres: ["Romance"],
      viewed: false,
      img: "https://picsum.photos/id/122/300/400",
      createdAt: "2022-10-16T20:51:45.249Z",
    },
    {
      id: "4",
      title: "Movie 4",
      genres: ["Test"],
      viewed: false,
      img: "https://picsum.photos/id/168/300/400",
      createdAt: "2022-10-16T20:52:14.784Z",
    },
    {
      id: "5",
      title: "Movie 5",
      genres: ["Random"],
      viewed: false,
      img: "https://picsum.photos/id/149/300/400",
      createdAt: "2022-10-16T20:52:36.529Z",
    },
  ],
};

export const singleMovieData = {
  id: "1",
  title: "Movie 1",
  genres: ["Horror", "Comedy", "Romance"],
  viewed: false,
  img: "https://picsum.photos/id/149/300/400",
  createdAt: "2022-10-16T20:52:36.529Z",
};

export const activeFiltersData = {
  activeFilters: [],
  search: "",
};

export const activeLoadingData = {
  value: false,
};

export const emptyStoreData = {
  movies: { viewedMovies: [], unViewedMovies: [] },
  filters: { activeFilters: [], search: "" },
  loading: { value: false },
};

export const sampleFiltersData = ["Horror", "Comedy", "Romance"];

export const customStoreData = (movies, activeFilters, loading) => {
  return {
    movies: movies ? moviesData : emptyStoreData.movies,
    filters: activeFilters ? activeFiltersData : emptyStoreData.filters,
    loading: loading ? activeLoadingData : emptyStoreData.loading,
  };
};
