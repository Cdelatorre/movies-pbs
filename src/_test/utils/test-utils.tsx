import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
