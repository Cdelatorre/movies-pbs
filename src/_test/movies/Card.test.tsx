import { fireEvent, screen } from "@testing-library/react";
import MovieCard from "../../components/movies/card/index";
import { singleMovieData as movie } from "../utils/mocks/mock";
import { renderWithProviders } from "../utils/test-utils";

describe("Renders MovieCard component correctly", () => {
  test("Renders default MovieCard and each component correctly", () => {
    renderWithProviders(<MovieCard {...movie} />);

    expect(screen.getByTestId(`movie-card-${movie.id}`)).toBeInTheDocument();
    expect(
      screen.getByRole("heading", {
        name: movie.title,
      })
    ).toBeInTheDocument();
    movie.genres.forEach((genre) => {
      expect(screen.getByTestId(`genre-tab-${genre}`)).toBeInTheDocument();
    });
  });

  test("Renders viewed state if movie has viewed prop as true", async () => {
    renderWithProviders(<MovieCard {...movie} viewed />);

    const movieCard = await screen.findByTestId(`movie-card-${movie.id}`);

    expect(movieCard).toHaveClass("viewed-card");
  });

  test("Change title tag for input to edit if edit button is checked", () => {
    renderWithProviders(<MovieCard {...movie} viewed />);

    const editButton = screen.getByTestId(`edit-movie-btn-${movie.id}`);

    expect(screen.queryByDisplayValue(movie.title)).toBeNull();
    expect(
      screen.getByRole("heading", {
        name: movie.title,
      })
    ).toBeInTheDocument();

    fireEvent.click(editButton);

    expect(
      screen.queryByRole("heading", {
        name: movie.title,
      })
    ).toBeNull();
    expect(screen.getByDisplayValue(movie.title)).toBeInTheDocument();
  });
});
