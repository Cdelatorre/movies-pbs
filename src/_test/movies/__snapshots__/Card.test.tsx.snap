// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders MovieCard component correctly Renders default MovieCard and each component correctly 1`] = `
<div
  class="movie-card card border-2 border-dark unViewed-card"
  data-testid="movie-card-1"
  id="1"
>
  <div
    class="card-img-top-container"
  >
    <div
      class="card-img-custom"
      style="background-image: url(https://picsum.photos/id/149/300/400);"
    />
    <div
      class="action-card-btn bg-dark viewed text-primary"
      data-testid="toggle-viewed-btn-1"
    >
      <i
        class="far fa-eye-slash text-warning"
      />
    </div>
    <div
      class="action-card-btn bg-dark delete text-danger"
      data-testid="delete-movie-btn-1"
    >
      <i
        class="far fa-trash-alt"
      />
    </div>
    <div
      class="action-card-btn bg-dark edit text-secondary"
      data-testid="edit-movie-btn-1"
    >
      <i
        class="far fa-edit"
      />
    </div>
  </div>
  <div
    class="card-body bg-dark text-light pt-4"
  >
    <h3
      class="card-title text-start mt-2"
    >
      Movie 1
    </h3>
    <p
      class="card-description text-start"
    >
      Some quick example text to build on the card title and make up thebulk of the card's content.
    </p>
    <div
      class="genres d-flex"
    >
      <a
        aria-current="page"
        class="genre-tab border border-info text-light active"
        href="/?genre=Horror"
      >
        <span
          data-testid="genre-tab-Horror"
        >
          Horror
        </span>
      </a>
      <a
        aria-current="page"
        class="genre-tab border border-info text-light active"
        href="/?genre=Comedy"
      >
        <span
          data-testid="genre-tab-Comedy"
        >
          Comedy
        </span>
      </a>
      <a
        aria-current="page"
        class="genre-tab border border-info text-light active"
        href="/?genre=Romance"
      >
        <span
          data-testid="genre-tab-Romance"
        >
          Romance
        </span>
      </a>
    </div>
  </div>
</div>
`;
